### PR TITLE
Simplify set and requirements logic

### PR DESCRIPTION
**Pull request by ellistarn**
**1. Issue, if available:**

**2. Description of changes:**

```
karpenter-6cf9b544fb-wn8sl controller 2022-06-06T21:01:48.488Z	ERROR	controller	Scheduling pod, incompatible with provisioner "default", incompatible requirements, for key karpenter.sh/capacity-type, [spot] not in [on-demand]	{"commit": "08998cb", "pod": "default/inflate-6776857d54-dmgn5"}
```
```
karpenter-6cf9b544fb-wn8sl controller 2022-06-06T21:05:08.529Z	ERROR	controller	Scheduling pod, incompatible with provisioner "default", no instance type satisfied resources {"cpu":"1","pods":"1"} and requirements karpenter.sh/provisioner-name In [default], karpenter.sh/capacity-type In [invalid], kubernetes.io/arch In [amd64], eks.amazonaws.com/nodegroup DoesNotExist []	{"commit": "08998cb", "pod": "default/inflate-57f4ccfb9f-fzmx6"}
```
```
karpenter-bd4dcc8b6-fht89 controller 2022-06-06T22:11:09.683Z	ERROR	controller	Scheduling pod, incompatible with provisioner "default", incompatible requirements, key foo does not have known values	{"commit": "3d0434a", "pod": "default/inflate-8c89655b-fqql4"}
```

**3. How was this change tested?**

**4. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cloudprovider/aws/instancetype.go [hunk 1]
```diff
@@ -117,29 +117,31 @@ func (i *InstanceType) computeRequirements() scheduling.Requirements {
 		v1.LabelArchStable:         sets.NewSet(i.architecture()),
 		v1.LabelOSStable:           sets.NewSet(v1alpha5.OperatingSystemLinux),
 		v1.LabelTopologyZone:       sets.NewSet(lo.Map(i.Offerings(), func(o cloudprovider.Offering, _ int) string { return o.Zone })...),
+		// Well Known to Karpenter
 		v1alpha5.LabelCapacityType: sets.NewSet(lo.Map(i.Offerings(), func(o cloudprovider.Offering, _ int) string { return o.CapacityType })...),
-		// Resources
-		v1alpha1.InstanceCPULabelKey:    sets.NewSet(fmt.Sprint(aws.Int64Value(i.VCpuInfo.DefaultVCpus))),
-		v1alpha1.InstanceMemoryLabelKey: sets.NewSet(fmt.Sprint(aws.Int64Value(i.MemoryInfo.SizeInMiB))),
+		// Well Known to AWS
+		v1alpha1.InstanceCPULabelKey:             sets.NewSet(fmt.Sprint(aws.Int64Value(i.VCpuInfo.DefaultVCpus))),
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/instancetype.go:123]: would be nice for these to match the upstream labels and our capacity type label:

```
v1alpha1.LabelInstanceCPU
v1alpha1.LabelInstanceMemory
...
```

File: pkg/scheduling/requirements.go [hunk 2]
```diff
@@ -131,21 +130,23 @@ func (r Requirements) Get(key string) sets.Set {
...
+	// Custom Labels must intersect, but if not defined are denied.
+	for key := range requirements.Keys().Difference(v1alpha5.WellKnownLabels) {
+		if operator := requirements.Get(key).Type(); r.Has(key) || operator == v1.NodeSelectorOpNotIn || operator == v1.NodeSelectorOpDoesNotExist {
+			continue
+		}
+		errs = multierr.Append(errs, fmt.Errorf("%s not in %s, key %s", requirements.Get(key), r.Get(key), key))
+	}
+	// Well Known Labels must intersect, but if not defined, are allowed.
+	errs = multierr.Append(errs, r.Intersects(requirements))
 	return errs
```
Comment (bwagner5) [inline pkg/scheduling/requirements.go:142]:
```suggestion
	return multierr.Append(errs, r.Intersects(requirements))
```

File: pkg/cloudprovider/aws/instancetype.go (hunk 1 above)
Comment (ellistarn) [reply pkg/cloudprovider/aws/instancetype.go:123]: Great suggestion!

File: pkg/scheduling/requirements.go [hunk 3]
```diff
@@ -131,21 +130,22 @@ func (r Requirements) Get(key string) sets.Set {
...
-	for key := range requirements {
-		if !keys.Has(key) {
+	// Custom Labels must intersect, but if not defined are denied.
+	for key := range requirements.Keys().Difference(v1alpha5.WellKnownLabels) {
+		if operator := requirements.Get(key).Type(); r.Has(key) || operator == v1.NodeSelectorOpNotIn || operator == v1.NodeSelectorOpDoesNotExist {
 			continue
 		}
-		existing := r[key]
-		incoming := requirements[key]
+		errs = multierr.Append(errs, fmt.Errorf("%s not in %s, key %s", requirements.Get(key), r.Get(key), key))
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/scheduling/requirements.go:138]: Here's what it looks like: `[us-west-2a] not in [us-west-2b], key topology.kubernetes.io/zone`. What's wrong with this to you?

File: pkg/cloudprovider/aws/instancetype.go [hunk 4]
```diff
@@ -117,29 +117,31 @@ func (i *InstanceType) computeRequirements() scheduling.Requirements {
...
 	// Instance Type Labels
 	instanceTypeParts := strings.Split(aws.StringValue(i.InstanceType), ".")
 	if len(instanceTypeParts) == 2 {
-		requirements.Add(scheduling.Requirements{
-			v1alpha1.InstanceFamilyLabelKey: sets.NewSet(instanceTypeParts[0]),
-			v1alpha1.InstanceSizeLabelKey:   sets.NewSet(instanceTypeParts[1]),
-		})
+		requirements[v1alpha1.LabelInstanceFamily].Insert(instanceTypeParts[0])
+		requirements[v1alpha1.LabelInstanceSize].Insert(instanceTypeParts[1])
 	}
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/cloudprovider/aws/instancetype.go:137]: No, we don't want to block provisioning if AWS launches a new instance type that doesn't conform. I could potentially see logging an error, but I'm wary about wiring context through. This only fails if EC2 launches a new instance type of a new format, and we simply wouldn't enable selection by family or size if this is the case.

**Review by njtran — APPROVED** [review]

**Pull request merged**